x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/logical/join/JoinTypes.java

```diff
@@ -0,0 +1,145 @@
+            this.id = (byte) id;
+            this.name = name;
+        }
+
+        @Override
+        public String joinName() {
+            return name;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeByte(id);
+        }
+    }
+
+    /**
+     * Join type for the USING clause - shorthand for defining an equi-join (equality join meaning the condition checks if columns across
+     * each side of the join are equal).
+     * One important difference is that the USING clause returns the join column only once, at the beginning of the result set.
+     */
+    public static class UsingJoinType implements JoinType {
+        private final List<Attribute> columns;
+        private final JoinType coreJoin;
+
+        public UsingJoinType(JoinType coreJoin, List<Attribute> columns) {
+            this.columns = columns;
+            this.coreJoin = coreJoin;
+        }
+
+        @Override
+        public String joinName() {
+            return coreJoin.joinName() + " USING " + columns.toString();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            throw new IllegalArgumentException("USING join type should not be serialized due to being rewritten");
+        }
+
+        public JoinType coreJoin() {
+            return coreJoin;
+        }
+
+        public List<Attribute> columns() {
+            return columns;
+        }
+
+        @Override
+        public boolean resolved() {
+            return Resolvables.resolved(columns);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(columns, coreJoin);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            UsingJoinType that = (UsingJoinType) o;
+            return Objects.equals(columns, that.columns) && coreJoin == that.coreJoin;
+        }
+
+        @Override
+        public String toString() {
+            return joinName();
+        }
+    }
+
+    /**
+     * Private class so it doesn't get used yet it is defined to showcase why the join type was defined as an interface instead of a simpler
+     * enum.
+     */
+    private abstract static class NaturalJoinType implements JoinType {
+
+        private final JoinType joinType;
+
+        private NaturalJoinType(JoinType joinType) {
+            this.joinType = joinType;
+        }
+
+        @Override
+        public String joinName() {
+            return "NATURAL " + joinType.joinName();
+        }
+    }
+
+    public static JoinType readFrom(StreamInput in) throws IOException {
+        byte id = in.readByte();
+        return switch (id) {
+            case 1 -> INNER;
+            case 2 -> LEFT;
+            case 3 -> RIGHT;
+            case 4 -> FULL;
+            case 5 -> CROSS;
```

Comment:
nit: maybe this can be made a little more robust by directly using the `CoreJoinType.id`s here, e.g. by iterating over the enum's values and throwing them into a map with ids (obtained from each enum element directly) as key. Would avoid repeating the assignment between enum element and id here.